rust/vapid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+                ),
+            ].iter()
+                .cloned()
+                .collect();
+        reply
+    }
+
+    #[test]
+    fn test_sign() {
+        let key = Key::generate();
+        let sub_val = serde_json::Value::from(String::from("mailto:mail@example.com"));
+
+        let mut claims: HashMap<String, serde_json::Value> = HashMap::new();
+        claims.insert(String::from("sub"), sub_val.clone());
+        let result = sign(key, &mut claims).unwrap();
+        let vresult = result.clone();
+
+        // println!("{}", result);
+
+        assert!(result.starts_with("Authorization: "));
+        assert!(result.contains(" vapid "));
+
+        // tear apart the auth token for the happy bits
+        let token = result.split(" ").nth(2).unwrap();
+        let sub_parts: Vec<&str> = token.split(",").collect();
+        let mut auth_parts: HashMap<String, String> = HashMap::new();
+        for kvi in &sub_parts {
+            let kv: Vec<String> = kvi.splitn(2, "=").map(|x| String::from(x)).collect();
+            auth_parts.insert(kv[0].clone(), kv[1].clone());
+        }
+        assert!(auth_parts.contains_key("t"));
+        assert!(auth_parts.contains_key("k"));
+
+        // now tear apart the token
+        let token: Vec<&str> = auth_parts.get("t").unwrap().split(".").collect();
+        assert_eq!(token.len(), 3);
+
+        let content = String::from_utf8(
+            base64::decode_config(token[0], base64::URL_SAFE_NO_PAD).unwrap(),
+        ).unwrap();
+        let items: HashMap<String, String> = serde_json::from_str(&content).unwrap();
+        assert!(items.contains_key("typ"));
+        assert!(items.contains_key("alg"));
+
+        let content: String = String::from_utf8(
+            base64::decode_config(token[1], base64::URL_SAFE_NO_PAD).unwrap(),
+        ).unwrap();
+        let items: HashMap<String, serde_json::Value> = serde_json::from_str(&content).unwrap();
+
+        assert!(items.contains_key("exp"));
+        assert!(items.contains_key("sub"));
+        assert!(items.get("sub") == Some(&sub_val));
+
+        // And verify that the signature works.
+        // we do integration verify in `test_verify`
+        verify(vresult).expect("Signed claims failed to self verify");
+
+    }
+
+    // TODO: Test fail cases, verification, values, integration
+
+    #[test]
+    fn test_verify() {
+        // Integration test with externally generated Authorization header.
+        let test_header = [
+            "Authorization: vapid t=eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJhdWQiOiJodHRwcz",
+            "ovL3B1c2guc2VydmljZXMubW96aWxsYS5jb20iLCJleHAiOiIxNDYzMDAxMzQwIiwic3ViIjoibWFp",
+            "bHRvOmFkbWluQGV4YW1wbGUuY29tIn0.4ZiULZaqZ8_7Cf2UYu7KO3eGaqZL5d4RZ6pwBvR0rcmTho",
+            "4WryVuZLfN-iMsHJ6Oc-4hkEZsMj8_32sXYSvTyg,k=BPD3F0hvy3Df69tjqRBN0ad08WH2nfaaxnp",
+            "kuIO6BV9Pa7p8xA8GauX0R_S-D-k82kcTNsCiJ6ML-zJisBpyybs",
```

Comment:
You can also express this I think with multiline strings in Rust if you'd like:

```rust
let a = "a \
    b \
c \
      d";

assert_eq!(a, "a b c d");
```

In that a `\` followed by a newline in the string literal will trim all leading whitespace up to the next part of the string. I believe this'd be equivalent:

```rust
let test_header = "\
            Authorization: vapid t=eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJhdWQiOiJodHRwcz\
            ovL3B1c2guc2VydmljZXMubW96aWxsYS5jb20iLCJleHAiOiIxNDYzMDAxMzQwIiwic3ViIjoibWFp\
            bHRvOmFkbWluQGV4YW1wbGUuY29tIn0.4ZiULZaqZ8_7Cf2UYu7KO3eGaqZL5d4RZ6pwBvR0rcmTho\
            4WryVuZLfN-iMsHJ6Oc-4hkEZsMj8_32sXYSvTyg,k=BPD3F0hvy3Df69tjqRBN0ad08WH2nfaaxnp\
            kuIO6BV9Pa7p8xA8GauX0R_S-D-k82kcTNsCiJ6ML-zJisBpyybs\
";

Reply:
Huh, cool. I'll investigate and clean that up. (or possibly let `rustfmt` take a swing at it.)